src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,103 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int currentCapacity;
+    private T[] listArray;
+    private int size;
+
+    public ArrayList() {
+        this.listArray = (T[]) new Object[DEFAULT_CAPACITY];
+        size = 0;
+        currentCapacity = DEFAULT_CAPACITY;
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
+        if ((size == 0 && index == 0) || (index >= 0 && index <= size)) {
+            if (size == currentCapacity) {
+                currentCapacity = currentCapacity + (currentCapacity >> 1);
+                T[] temp = (T[]) new Object[currentCapacity];
+                System.arraycopy(listArray, 0, temp, 0, index);
+                System.arraycopy(listArray, index, temp, index + 1, size - index);
+                listArray = temp;
+            } else {
+                if (size > 0) {
+                    System.arraycopy(listArray, index, listArray, index + 1, size - index);
+                }
+            }
+            listArray[index] = value;
+            size++;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Can't add to index " + index);
+        }
 
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list == null) {
+            return;
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= 0 && index < size) {
+            return listArray[index];
+        }
+        throw new ArrayListIndexOutOfBoundsException("Can't get to index " + index);
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= 0 && index < size) {
+            listArray[index] = value;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Can't set to index " + index);
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (size == 0 || index < 0 || index >= size) {
```

Comment:
I see again a repeated code. Try to join size and index checks from get, set, remove(int index) methods in a separate private method.